### PR TITLE
Improvements in run-length logic and tests

### LINKED ISSUE #657: Many run length methods are not respecting their `dim` argument
* xclim version: master

Methods in `run_length` accept a `dim` argument to control over which dimensions the runs are sought for. However, a lot of them have hardcoded uses of `'time'`, instead of using that `dim` argument.

In the vast majority of cases, this is not an issue, but it happens. In another project, I used `first_run` over a spatial dimension, but had to code a workaround.

### LINKED ISSUE #717: Run length - dask-version not accepting integers.
The non-ufunc version of `run_length` (`xc.indices.run_length.rle`) doesn't accept integer arrays, while the `ufunc` one does (`xc.indices.run_length.rle_1d`).

On one side, this makes sense because the latter does more than the former : it finds all runs of _different values_. Not only runs of True values. However, this was the cause of a bug in #716 that was really hard to pin down. Passing integers (0 and 1) to `rle` results in no runs found (not np.NaN, but 0 runs).

Suggestion : Cast the input to boolean in `rle`. 
Suggestion #2 : Make the `ufunc_dim` argument controllable xclim-wide, through a `set_options` argument, facilitating testing and debug.

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #717 and fixes #657
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
- [ ] `bumpversion (major / minor / patch)` has been called on this branch
- [ ] Tags have been pushed (`git push --tags`)

* **What kind of change does this PR introduce?** 
    - Adds an option for controlling the version of the run length algorithms (ufunc vs broadcasting).
    - Uses that option to run all run length tests twice, once for each version.
    - Makes the algo version switch smarter by regrouping everything in a single function. I used the same logic than the Indicator class where "from_context" means the option is read from the options (that's the default).
    - Removes some repetive tests (I might have removed too many?)
    - Removes all hardcoded mentions of `"time"` in run length functions.
    - ~Tiny little, completely unrelated change in `calendar.py` to silence a DeprecationWarning.~
    - Slightly related : added `set_options` to the API page of the documentation

* **Does this PR introduce a breaking change?** 
No.

* **Other information**:
Right now, the default is to switch to the "broadcasting" version for: 1) large arrays and 2) dask arrays. Actually, the "ufunc" version does work with dask arrays, just not when the main dimension ('time') is chunked. In reality, the other version also shouldn't really work in that case, but we implemented a workaround that rechunks the data on-the-fly.

I suggest that we leave it as-is for now, but consider spending some time in real benchmarking. If it is more efficient (and I think it is), then the ufunc version could be used for single-chunk dask arrays too. Also, we could get a better idea of the optimal value for `npts_opt`. Finally, it could give us ideas for #505.

### CODE REVIEW

File: xclim/core/calendar.py [hunk 1]
```diff
@@ -340,7 +340,7 @@ def _convert_datetime(
     datetime: Union[pydt.datetime, cftime.datetime],
     new_doy: Optional[Union[float, int]] = None,
     calendar: str = "default",
-) -> Union[cftime.datetime, pydt.datetime, np.int]:
+) -> Union[cftime.datetime, pydt.datetime, int]:
```
Comment (Zeitsperre) [inline xclim/core/calendar.py:343]: I addressed this in #716 (and np.nan is technically a float!)

Comment (aulemahal) [reply xclim/core/calendar.py:343]: Oh woups. I'll revert this. Thanks!

**Comment by aulemahal** [comment]
For context, dear code readers, please notice the use of a super-dooper-fancy fixture at the head of `test_run_length.py`.  Because it has `scope='module'` and `autouse=True`, it gets executed for ALL tests. So they ALL run twice.

File: HISTORY.rst [hunk 2]
```diff
@@ -9,10 +9,11 @@ New features and enhancements
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Rewrite of nearly all adjustment methods in ``sdba``, with use of ``xr.map_blocks`` to improve scalability with dask. Rewrite of some parts of the algorithms with numba-accelerated code.
 * "GFWED" specifics for fire weather computation implemented back into the FWI module. Outputs are within 3% of GFWED data.
+* Addition of the `run_length_ufunc` option to control which run length algorithm gets run. Defaults stay the same (automatic switch dependent of the input array).
 
 Breaking changes
 ~~~~~~~~~~~~~~~~
-* The `tropical_nights` indice is being deprecated in favour of `tn_days_above` with `thresh="20 degC"`. The indicator remains valid, now wrapping this new indice.
+* The `tropical_nights` indice is being deprecated in favour of `tn_days_above` with ``thresh="20 degC"``. The indicator remains valid, now wrapping this new indice.
```
Comment (huard) [inline HISTORY.rst:16]: Did `set_options` remove the update method from the public interface ?

File: HISTORY.rst [hunk 3]
```diff
@@ -9,10 +9,11 @@ New features and enhancements
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Rewrite of nearly all adjustment methods in ``sdba``, with use of ``xr.map_blocks`` to improve scalability with dask. Rewrite of some parts of the algorithms with numba-accelerated code.
 * "GFWED" specifics for fire weather computation implemented back into the FWI module. Outputs are within 3% of GFWED data.
+* Addition of the `run_length_ufunc` option to control which run length algorithm gets run. Defaults stay the same (automatic switch dependent of the input array).
```
Comment (huard) [inline HISTORY.rst:12]: Explain the criteria for the automated switch.

**Review by huard — APPROVED** [review]

File: HISTORY.rst (hunk 2 above)
Comment (aulemahal) [reply HISTORY.rst:16]: Yes, but it was never intended to be public I think. Users are not supposed to use the `set_options` instance.

**Pull request merged**